pkg/engine/executor/merge_test.go

```diff
@@ -2,96 +2,136 @@ package executor
+		// Each pipeline has 2-4 records with unique identifiers
+		numRecords := 2 + (i % n) // 2, 3, or 4 records per pipeline
+		rows := make([]arrowtest.Rows, numRecords)
+
+		for j := range numRecords {
+			// Each record has 3-5 rows
+			numRows := 3 + (j % n) // 3, 4, or 5 rows per record
+			recordRows := make(arrowtest.Rows, numRows)
+
+			for k := range numRows {
+				recordRows[k] = map[string]any{
+					"timestamp": int64(i*100 + j*10 + k),
+					"message":   fmt.Sprintf("input=%d record=%d row=%d", i, j, k),
+				}
+			}
+			rows[j] = recordRows
 		}
 
-		rec, _ := m.Value()
-		defer rec.Release()
-
-		rows, err := arrowtest.RecordRows(rec)
-		require.NoError(t, err)
-
-		got = append(got, rows)
+		// Accumulate expected rows
+		expectedRows = append(expectedRows, rows...)
+		inputRows[i] = rows
 	}
 
-	require.ErrorIs(t, err, EOF)
-
-	require.Equal(t, append(rowsInput1, rowsInput2...), got)
+	// Test different concurrency settings
+	//  0 ... no prefetching of next input / only current input is prefetched
+	//  1 ... next input is prefetched
+	//  3 ... number of prefetched inputs is equal to the number of inputs
+	//  5 ... number of prefetched inputs is greater than the number of inputs
+	// -1 ... unlimited/all inputs are prefetched
+	for _, maxPrefetch := range []int{0, 1, n, 5, -1} {
+
+		t.Run(fmt.Sprintf("context=full/maxPrefetch=%d", maxPrefetch), func(t *testing.T) {
+			// Create fresh inputs using the pre-generated data
+			inputs := make([]Pipeline, n)
+			for i := range len(inputs) {
+				inputs[i] = NewArrowtestPipeline(alloc, schema, inputRows[i]...)
+			}
+
+			m, err := newMergePipeline(inputs, maxPrefetch)
+			require.NoError(t, err)
+			defer m.Close()
+
+			ctx := t.Context()
+			var actualRows []arrowtest.Rows
+
+			// Read all records from the merge pipeline
+			for {
+				err := m.Read(ctx)
+				if errors.Is(err, EOF) {
+					t.Log("stop reading from pipeline:", err)
+					break
+				}
+				require.NoError(t, err, "Unexpected error during read")
+
+				rec, _ := m.Value()
+				defer rec.Release()
+
+				rows, err := arrowtest.RecordRows(rec)
+				require.NoError(t, err)
+
+				actualRows = append(actualRows, rows)
+			}
+
+			// Compare actual vs expected rows
+			// Order of processed inputs must stay the same
+			require.Equal(t, expectedRows, actualRows)
+		})
+
+		t.Run(fmt.Sprintf("context=canceled/maxPrefetch=%d", maxPrefetch), func(t *testing.T) {
+			// Create fresh inputs using the pre-generated data
+			inputs := make([]Pipeline, n)
+			for i := range len(inputs) {
+				inputs[i] = NewArrowtestPipeline(alloc, schema, inputRows[i]...)
+			}
+
+			m, err := newMergePipeline(inputs, maxPrefetch)
+			require.NoError(t, err)
+			defer m.Close()
+
+			ctx := t.Context()
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// Read all records from the merge pipeline until context is canceled
+			iter := -1
+			for {
+				iter++
+				if iter == 3 {
+					cancel()
+				}
```

Comment:
This bit feels a little magical and hard to follow for me. Does 3 come from the setting of `n`? Can we update to something like this? 

```go
for iter := 0; true; iter++ {
  if iter == n {
    cancel() 
  }
}
```

Reply:
I guess I could count the rows generated and stop once half of them are consumed.